(Samples) Transfer in (#231)
OfficeDev/office-scripts-docs: docs/resources/samples/move-rows-across-tables.md

---
title: 'Move rows across tables using Office Scripts'
description: 'Learn how to move rows across tables by saving filters, then processing and reapplying the filters.'
ms.date: 03/30/2021
localization_priority: Normal
---

# Move rows across tables by saving filters, then processing and reapplying the filters

> This script achieves the following:

Reviewer: Sounds less formal:
```suggestion
This script does the following:
```

* Select rows from the source table where the value in a column is equal to _some value_.
* Move all selected rows into another (target) table on another worksheet.

> * Reapply the relevant filters on the source table.

Reviewer: I'm not sure the verb tense on these is right. Shouldn't it be "Selects", "Moves", and "Reapplies"?

![Screenshots of workbook before and after](../../images/table-filter-before-after.png)

> There are two script variants in this solution. The main difference is in the way the rows are selected.

Reviewer: Suggestion for brevity
```suggestion
There are two scripts in this solution. The main difference is how the rows are selected.
```

* In the [first variant](#office-scripts-sample-code-move-rows-using-table-filter), the rows are selected by applying the table filter and reading the visible range.
* In the [second](#office-scripts-sample-code-move-rows-using-range-values), the rows are selected by reading the values and extracting the row values.

## Sample Excel file

> Download the file <a href="input-table-filters.xlsx">input-table-filters.xlsx</a> used in this solution to try it out yourself!

Reviewer: @ElizabethSamuel-MSFT How do you feel about exclamation points?

## Office Scripts sample code: Move rows using table filter

```TypeScript
function main(workbook: ExcelScript.Workbook) {

  // Update the table names, column name, and look up value as needed.
  const TargetTableName = 'Table1';
  const SourceTableName = 'Table2';
  const NameOfColumnToFilterOn = 'Category';
  const ValueToFilterOn = 'Clothing';

  // Get the table objects.
  let targetTable = workbook.getTable(TargetTableName);
  let sourceTable = workbook.getTable(SourceTableName);

>   // If you don't know the table names, you can fetch first table on a given worksheet name.

Reviewer: ```suggestion
  // If you don't know the table names, you can use the following code instead to fetch first table on a given worksheet name.
```

---
title: 'Move rows across tables in Office Scripts'
description: 'Learn how to move rows across tables by saving filters, then processing and reapplying the filters.'
ms.date: 03/30/2021
localization_priority: Normal
---

# Move rows across tables by saving filters, then processing and reapplying the filters

This script achieves the following:

* Select rows from the source table where the value in a column is equal to _some value_.
* Move all selected rows into another (target) table on another worksheet.
* Reapply the relevant filters on the source table.

![Screenshots of workbook before and after](../../images/table-filter-before-after.png)

## Solution

There are two script variants. The main difference is in the way the rows are selected.

* In the [first variant](#move-rows-using-table-filter), the rows are selected by applying the table filter and reading the visible range.
* In the [second](#move-rows-using-range-values), the rows are selected by reading the values and extracting the row values.

## Sample Excel file

Download the file <a href="input-table-filters.xlsx">input-table-filters.xlsx</a> used in this solution to try it out yourself!

> ## Office Scripts

Reviewer 2: Please see suggestion in previous file removing unnecessary navigation and heading levels.

## Office Scripts

The solution contains two Office Scripts:

1. [Move Rows Using Table Filter](#move-rows-using-table-filter)
1. [Move Rows Using Range Values](#move-rows-using-range-values)

### Move rows using table filter

```ts
function main(workbook: ExcelScript.Workbook) {

  // Update the table names, column name, and look-up value as needed.
  const TargetTableName = 'Table1';
  const SourceTableName = 'Table2';
  const NameOfColumnToFilterOn = 'Category';
  const ValueToFilterOn = 'Clothing';

  // Get the table objects.
  let targetTable = workbook.getTable(TargetTableName);
  let sourceTable = workbook.getTable(SourceTableName);
  // If you don't know the table names, you can fetch first table on a given worksheet name.
  // let targetTable = workbook.getWorksheet('Sheet1').getTables()[0];
  // let sourceTable = workbook.getWorksheet('Sheet2').getTables()[0];

  
  if (!targetTable || !sourceTable) {
    console.log(`Required tables missing - Check that both source (${TargetTableName}) and target (${SourceTableName}) tables are present before running the script.`);
    return;

>   }

Reviewer: There's a lot of error handling and input checking here. Do we really need that for the sample? I think it detracts from what we're trying to show.

Author: Tracking with [feature](https://office.visualstudio.com/OC/_workitems/edit/4924924)

  // Save all of the filter criteria.
  // Initialize with an empty object to hold the filter criteria.
  const tableFilters = {};
  // For each table column, collect the filter criteria.
  sourceTable.getColumns().forEach((column) => {
    let colFilterCriteria = column.getFilter().getCriteria();
    if (colFilterCriteria) {

>       // If we don't remove these two keys, the API fails for some reason. So, remove these.

Reviewer: We should know why this happens.

Author: Tracking with [feature](https://office.visualstudio.com/OC/_workitems/edit/4924924)

Reviewer 3: The issue has been reported and is being addressed by Excel API team.
